src/internal/target-runner.csx

```diff
@@ -60,6 +57,44 @@ public static class SimpleTargetsTargetRunner
         }
     }
 
+    private static void ValidateDependencies(IDictionary<string, Target> targets)
+    {
+        var missingDependencies = new SortedDictionary<string, ISet<string>>();
+
+        foreach (var targetEntry in targets)
+        {
+            foreach (var dependencyName in targetEntry.Value.Dependencies)
+            {
+                if (!targets.ContainsKey(dependencyName))
+                {
+                    var missingDependenciesForTarget = GetOrAddEmpty(missingDependencies, dependencyName);
+                    missingDependenciesForTarget.Add(targetEntry.Key);
+                }
+            }
+        }
+
+        if (!missingDependencies.Any())
+        {
+            return;
+        }
+
+        var manyDependencies = missingDependencies.Count() > 1;
+        var message = new StringBuilder($"Missing {(manyDependencies ? "dependencies" : "dependency")} detected:");
+        foreach (var missingDependencyEntry in missingDependencies)
+        {
+            var targetsCsv = string.Join(", ", missingDependencyEntry.Value.Select(target => $"\"{target.Replace("\"", "\"\"")}\""));
+            message.Append($@"\r\n  ""{missingDependencyEntry.Key.Replace("\"", "\"\"")}"", required by {targetsCsv}");
+        }
+
+        throw new Exception(message.ToString());
+    }
+
+    private static ISet<string> GetOrAddEmpty(IDictionary<string, ISet<string>> dictionary, string key)
+    {
+        ISet<string> set;
+        return (dictionary.TryGetValue(key, out set)) ? set : dictionary[key] = new SortedSet<string>();
+    }

```

Comment:
(Pure refactoring, not essential.)

Using `out var`, this can be collapsed to an expression:

```C#
private static ISet<string> GetOrAddEmpty(IDictionary<string, ISet<string>> dictionary, string key) =>
    (dictionary.TryGetValue(key, out var set)) ? set : dictionary[key] = new SortedSet<string>();
```

or even, this can be in-lined back at the call site (shown together with some other abbreviations):

```C#
foreach (var dependencyName in targetEntry.Value.Dependencies
    .Where(dependencyName => !targets.ContainsKey(dependencyName)))
{
    (missingDependencies.TryGetValue(dependencyName, out var set)
            ? set
            : missingDependencies[dependencyName] = new SortedSet<string>())
        .Add(targetEntry.Key);
}
```